packages/flutter/lib/src/widgets/scroll_position.dart

```diff
@@ -164,12 +162,12 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
   double _impliedVelocity = 0;
 
   @override
-  double get pixels => _pixels;
-  double _pixels;
+  double? get pixels => _pixels;
+  double? _pixels;
```

Comment:
this should be non-nullable unless there's a really good reason it has to be nullable...

Reply:
There are some specific handling when pixels is null. (Search `pixels == null`)

Reply:
Those are all about the initial value. At a minimum let's make `pixels` non-nullable and add a `hasPixels` getter that returns whether `_pixels` is non-null. The `pixels` getter itself is almost always assumed to return non-null.

Reply:
`hasPixels` getter added.

Reply:
Can we keep hasPixels to ScrollPosition rather than putting it on VIewportOffset itself?

Reply:
`pixels` is currently used in `VIewportOffset.debugFillDescription` and we should guard its usage to avoid LazyInitializationError. That why I added `VIewportOffset.hasPixels`.
Alternativelly we could move this `description.add('offset: ${pixels.toStringAsFixed(1)}');` to every child having the concret implementation of `double get pixels`.

```diff
diff --git a/packages/flutter/lib/src/rendering/viewport_offset.dart b/packages/flutter/lib/src/rendering/viewport_offset.dart
index 3e986f8640..435cbfb182 100644
--- a/packages/flutter/lib/src/rendering/viewport_offset.dart
+++ b/packages/flutter/lib/src/rendering/viewport_offset.dart
@@ -94,9 +94,6 @@ abstract class ViewportOffset extends ChangeNotifier {
   /// the value changes due to [correctBy]).
   double get pixels;
 
-  /// Whether [pixels] property is available.
-  bool get hasPixels;
-
   /// Called when the viewport's extents are established.
   ///
   /// The argument is the dimension of the [RenderViewport] in the main axis
@@ -247,11 +244,7 @@ abstract class ViewportOffset extends ChangeNotifier {
   /// If you override this, make sure to start your method with a call to
   /// `super.debugFillDescription(description)`.
   @mustCallSuper
-  void debugFillDescription(List<String> description) {
-    if (hasPixels) {
-      description.add('offset: ${pixels.toStringAsFixed(1)}');
-    }
-  }
+  void debugFillDescription(List<String> description) { }
 }
 
 class _FixedViewportOffset extends ViewportOffset {
@@ -263,9 +256,6 @@ class _FixedViewportOffset extends ViewportOffset {
   @override
   double get pixels => _pixels;
 
-  @override
-  bool get hasPixels => true;
-
   @override
   bool applyViewportDimension(double viewportDimension) => true;
 
@@ -294,4 +284,10 @@ class _FixedViewportOffset extends ViewportOffset {
 
   @override
   bool get allowImplicitScrolling => false;
+
+  @override
+  void debugFillDescription(List<String> description) {
+    super.debugFillDescription(description);
+    description.add('offset: ${pixels.toStringAsFixed(1)}');
+  }
 }
diff --git a/packages/flutter/lib/src/widgets/scroll_position.dart b/packages/flutter/lib/src/widgets/scroll_position.dart
index 2537333b73..25864e3cea 100644
--- a/packages/flutter/lib/src/widgets/scroll_position.dart
+++ b/packages/flutter/lib/src/widgets/scroll_position.dart
@@ -901,6 +901,8 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
     if (debugLabel != null)
       description.add(debugLabel!);
     super.debugFillDescription(description);
+    if (hasPixels)
+      description.add('offset: ${pixels.toStringAsFixed(1)}');
     description.add('range: ${_minScrollExtent?.toStringAsFixed(1)}..${_maxScrollExtent?.toStringAsFixed(1)}');
     description.add('viewport: ${_viewportDimension?.toStringAsFixed(1)}');
   }
diff --git a/packages/flutter/test/rendering/slivers_block_test.dart b/packages/flutter/test/rendering/slivers_block_test.dart
index cdb61b8add..05bb823a61 100644
--- a/packages/flutter/test/rendering/slivers_block_test.dart
+++ b/packages/flutter/test/rendering/slivers_block_test.dart
@@ -79,9 +79,6 @@ class ViewportOffsetSpy extends ViewportOffset {
   @override
   double get pixels => _pixels;
 
-  @override
-  bool get hasPixels => true;
-
   bool corrected = false;
 
   @override
```

WDYT?

Reply:
Interesting trade-off. I think you were probably right the first time, ViewportOffset.hasPixels makes sense since I guess we're saying in the fillDescription that it is legitimate that it can sometimes be null.